File: Fibonacci/Fibonacci.cpp
#include <iostream>
#include <chrono>
#include <windows.h>
#include <psapi.h>

long long fibonacci_recursive(int n) {
    if (n <= 1) return n;
    return fibonacci_recursive(n-1) + fibonacci_recursive(n-2);
}

int main() {
    const int n = 35;
    
    // Measure initial memory
    PROCESS_MEMORY_COUNTERS memInfo;
    GetProcessMemoryInfo(GetCurrentProcess(), &memInfo, sizeof(memInfo));
    double mem_before = memInfo.PeakWorkingSetSize / (1024.0 * 1024.0); // MB
    
    // Compute Fibonacci with naive recursion
    auto start = std::chrono::high_resolution_clock::now();
    clock_t start_cpu = clock();
    long long result = fibonacci_recursive(n);
    clock_t end_cpu = clock();
    auto end = std::chrono::high_resolution_clock::now();
    
    // Measure final memory
    GetProcessMemoryInfo(GetCurrentProcess(), &memInfo, sizeof(memInfo));
    double mem_after = memInfo.PeakWorkingSetSize / (1024.0 * 1024.0); // MB
    double max_memory = std::max(mem_before, mem_after);
    
    // Output metrics
    std::chrono::duration<double> diff = end - start;
    std::cout << "Execution time: " << diff.count() << " seconds\n";
    std::cout << "CPU time: " << (end_cpu - start_cpu) / (double)CLOCKS_PER_SEC << " seconds\n";
    std::cout << "Max memory: " << max_memory << " MB\n";
    std::cout << "Fibonacci(" << n << ") = " << result << "\n";
    return 0;
}

File: Fibonacci/Fibonacci.java

import java.lang.management.*;

public class Fibonacci {
    public static long fibonacciRecursive(int n) {
        if (n <= 1) return n;
        return fibonacciRecursive(n-1) + fibonacciRecursive(n-2);
    }
    
    public static void main(String[] args) {
        int n = 35;
        
        // Measure initial memory
        Runtime runtime = Runtime.getRuntime();
        MemoryMXBean memoryBean = ManagementFactory.getMemoryMXBean();
        runtime.gc();
        long mem_before = runtime.totalMemory() - runtime.freeMemory();
        long non_heap_before = memoryBean.getNonHeapMemoryUsage().getUsed();
        
        // Compute Fibonacci with naive recursion
        ThreadMXBean bean = ManagementFactory.getThreadMXBean();
        long start_cpu = bean.isCurrentThreadCpuTimeSupported() ? bean.getCurrentThreadCpuTime() : 0;
        long start_time = System.nanoTime();
        long result = fibonacciRecursive(n);
        long end_time = System.nanoTime();
        long end_cpu = bean.isCurrentThreadCpuTimeSupported() ? bean.getCurrentThreadCpuTime() : 0;
        
        // Measure final memory
        runtime.gc();
        long mem_after = runtime.totalMemory() - runtime.freeMemory();
        long non_heap_after = memoryBean.getNonHeapMemoryUsage().getUsed();
        
        // Calculate memory usage
        long heap_memory = Math.max(mem_before, mem_after) / (1024 * 1024); // Heap in MB
        long non_heap_memory = Math.max(non_heap_before, non_heap_after) / (1024 * 1024); // Non-heap in MB
        long total_memory = (Math.max(mem_before, mem_after) + Math.max(non_heap_before, non_heap_after)) / (1024 * 1024); // Total in MB
        
        // Output metrics
        System.out.printf("Execution time: %.4f seconds\n", (end_time - start_time) / 1e9);
        System.out.printf("CPU time: %.4f seconds\n", (end_cpu - start_cpu) / 1e9);
        System.out.printf("Heap memory used: %d MB\n", heap_memory);
        System.out.printf("Non-heap memory used: %d MB\n", non_heap_memory);
        System.out.printf("Total memory used: %d MB\n", total_memory);
        System.out.println("Fibonacci(" + n + ") = " + result);
    }
}
